docs/content/guides/developer/advanced/randomness-onchain.mdx

```diff
@@ -20,76 +20,148 @@ Although `Random` is a shared object, it is inaccessible for mutable operations,
 
 :::
 
-Having access to random numbers is only one part of designing secure applications, you should also pay careful attention to how you use that randomness. To securely access randomness:
+Having access to random numbers is only one part of designing secure applications, you should also pay careful attention to how you use that randomness.
+To securely access randomness:
 
 - Define your function as (private) `entry`.
 - Prefer generating randomness using function-local `RandomGenerator`.
 - Make sure that the "unhappy path" of your function does not charge more gas than the "happy path".
 
+## Limited resources and `Random` dependent flows
+
+Developers must be aware that some resources that are available to transactions are limited.
+If a function that reads `Random` consumes more resources in an unhappy path flow than in a happy path flow, an attacker can use that difference to revert the transaction in the unhappy flow.
+
+Concretely, gas is such a resource.
+Consider the following vulnerable code:
+
+```move
+// Insecure implementation, do not use!
+entry fun insecure_play(r: &Random, payment: Coin<SUI>, ...) {
+  ...
+  let mut generator = new_generator(r, ctx);
+  let win = generator.generate_bool();
+  if (win) { // happy flow
+    ... cheap computation ...
+  } else {
+    ... very expensive computation ...
+  }
+}
+```
+
+Observe that the gas costs of a transaction that calls `insecure_play` depends on the value of `win`.
+An attacker could call this function with a gas budget that is sufficient for the happy flow but not the unhappy one, resulting in it either winning or reverting the transaction (but never losing the payment).
+
+:::warning
+
+Sui does not prevent this kind of attack, and developers must be aware of this subtlety when designing their contracts.
+
+:::
+
+
+Other limited resources per transaction that developers should consider are:
+
+- The number of new objects.
+- The number of objects that can be used (including dynamic fields).
+- Number of events emitted.
+- Number of UIDs generated, or deleted, or transferred.
+
+
+
+For many use cases this attack is not an issue, like when selecting a raffle winner, or lottery numbers, as the code running is independent of the randomness.
+However, in the cases where it can be problematic, you can consider one of the following:
+
+- Use two steps:
```

Comment:
The description here is necessarily a bit abstract - I think you should also link from here directly to the two-step example you wrote in code